core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java

```diff
@@ -401,6 +434,39 @@ private static void handleException(ClientRequestContext ctx, HttpRequestDuplica
         };
     }
 
+    private void redirect(ClientRequestContext ctx, ClientRequestContext derivedCtx,
+                          HttpRequestDuplicator rootReqDuplicator, CompletableFuture<HttpResponse> future,
+                          CompletableFuture<Void> returnedResWhenComplete, HttpResponse originalRes,
+                          Backoff backoff, RequestHeaders newHeaders) {
+        final HttpRequest originalReq = getOriginalRequest(ctx);
+        assert originalReq != null;
+        final String originalPath = originalReq.path();
+        final String newPath = newHeaders.path();
+        if (newPath.equals(originalPath) || !addRedirectPath(ctx, newPath)) {
+            final Set<String> paths = redirectPaths(ctx);
+            assert paths != null;
+            final RedirectLoopsException exception = new RedirectLoopsException(originalPath, paths);
+            abortResponse(originalRes, derivedCtx, exception);
+            handleException(ctx, rootReqDuplicator, future, exception, false);
+            return;
+        }
+        abortResponse(originalRes, derivedCtx, null);
+        resetTotalAttempts(ctx);
+        final long millisAfter = useRetryAfter ? getRetryAfterMillis(derivedCtx)
+                                               : 0; // nextDelay will be at least 0.
+        final long nextDelay = Math.max(backoff.nextDelayMillis(1), millisAfter);
+        final HttpRequestDuplicator newReqDuplicator;
+        if (rootReqDuplicator.headers().method() != newHeaders.method()) {
+            rootReqDuplicator.abort();
+            newReqDuplicator = HttpRequest.of(newHeaders).toDuplicator();
```

Comment:
Is it efficient if we introduce `EmptyBodyRequestDuplicator`?

Reply:
That sounds good. Let me add a TODO for that. 😄